src/ctia/flows/crud.clj

```diff
@@ -7,20 +7,35 @@
             [ctia.flows.hooks :as h]
             [ctia.events.obj-to-event :refer [to-create-event
                                               to-update-event
-                                              to-delete-event]]))
+                                              to-delete-event]]
+            [ring.util.http-response :as http-response]))
 
-(defn find-id
+(defn- find-id
   "Lookup an ID in a given entity.  Parse it, because it might be a
    URL, and return the short form ID.  Returns nil if the ID could not
    be found."
   [{id :id}]
   (when (seq id)
     (id/str->short-id id)))
 
-(defn make-id
+(defn- make-id
   [entity-type]
   (str (name entity-type) "-" (UUID/randomUUID)))
 
+(defn- find-entity-id [identity entity-type entity prev-entity]
+  (or (find-id prev-entity)
+      (when-let [entity-id (find-id entity)]
+        (when-not (auth/capable? identity :specify-id)
+          (throw (http-response/bad-request!
+                  {:error "Missing capability to specify entity ID"
+                   :entity entity})))
+        (if (id/valid-short-id? entity-id)
```

Comment:
@polygloton seems more consistent to use `when-not` twice here, instead of the combo of `when-not` and `if`

```clj
(defn- find-entity-id [identity entiaty-type entity prev-entity]
  (or (find-id prev-entity)
      (when-let [entity-id (find-id entity)]
        (when-not (auth/capable? identity :specify-id)
          (throw (http-response/bad-request!
                  {:error "Missing capability to specify entity ID"
                   :entity entity})))
        (when-not (id/valid-short-id? entity-id)
          (throw (http-response/bad-request!
                  {:error (format "Invalid entity ID: %s" entity-id)
                   :entity entity})))
        entity-id)
      (make-id entity-type)))
```